### PR TITLE
HADOOP-19218 Avoid DNS lookup while creating IPC Connection object

### PR DESCRIPTION
**Pull request by virajjasani**
### Description of PR
Jira: HADOOP-19218

### How was this patch tested?
Manual testing + UT

### For code changes:

- [X] Does the title or this PR starts with the corresponding JIRA issue id (e.g. 'HADOOP-17799. Your PR title ...')?
- [ ] Object storage: have the integration tests been executed and the endpoint declared according to the connector-specific documentation?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the `LICENSE`, `LICENSE-binary`, `NOTICE-binary` files?

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 21s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 2 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  38m  9s |  |  trunk passed  |
| -1 :x: |  compile  |   0m 45s | [/branch-compile-root-jdkUbuntu-11.0.23+9-post-Ubuntu-1ubuntu120.04.2.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6916/1/artifact/out/branch-compile-root-jdkUbuntu-11.0.23+9-post-Ubuntu-1ubuntu120.04.2.txt) |  root in trunk failed with JDK Ubuntu-11.0.23+9-post-Ubuntu-1ubuntu120.04.2.  |
| -1 :x: |  compile  |   0m 49s | [/branch-compile-root-jdkPrivateBuild-1.8.0_412-8u412-ga-1~20.04.1-b08.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6916/1/artifact/out/branch-compile-root-jdkPrivateBuild-1.8.0_412-8u412-ga-1~20.04.1-b08.txt) |  root in trunk failed with JDK Private Build-1.8.0_412-8u412-ga-1~20.04.1-b08.  |
| +1 :green_heart: |  checkstyle  |   0m 35s |  |  trunk passed  |
| -1 :x: |  mvnsite  |   0m 12s | [/branch-mvnsite-hadoop-common-project_hadoop-common.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6916/1/artifact/out/branch-mvnsite-hadoop-common-project_hadoop-common.txt) |  hadoop-common in trunk failed.  |
| -1 :x: |  javadoc  |   0m 23s | [/branch-javadoc-hadoop-common-project_hadoop-common-jdkUbuntu-11.0.23+9-post-Ubuntu-1ubuntu120.04.2.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6916/1/artifact/out/branch-javadoc-hadoop-common-project_hadoop-common-jdkUbuntu-11.0.23+9-post-Ubuntu-1ubuntu120.04.2.txt) |  hadoop-common in trunk failed with JDK Ubuntu-11.0.23+9-post-Ubuntu-1ubuntu120.04.2.  |
| -1 :x: |  javadoc  |   0m 20s | [/branch-javadoc-hadoop-common-project_hadoop-common-jdkPrivateBuild-1.8.0_412-8u412-ga-1~20.04.1-b08.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6916/1/artifact/out/branch-javadoc-hadoop-common-project_hadoop-common-jdkPrivateBuild-1.8.0_412-8u412-ga-1~20.04.1-b08.txt) |  hadoop-common in trunk failed with JDK Private Build-1.8.0_412-8u412-ga-1~20.04.1-b08.  |
| +1 :green_heart: |  spotbugs  |   1m 22s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  23m 46s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 33s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   9m 38s |  |  the patch passed with JDK Ubuntu-11.0.23+9-post-Ubuntu-1ubuntu120.04.2  |
| -1 :x: |  javac  |   9m 38s | [/results-compile-javac-root-jdkUbuntu-11.0.23+9-post-Ubuntu-1ubuntu120.04.2.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6916/1/artifact/out/results-compile-javac-root-jdkUbuntu-11.0.23+9-post-Ubuntu-1ubuntu120.04.2.txt) |  root-jdkUbuntu-11.0.23+9-post-Ubuntu-1ubuntu120.04.2 with JDK Ubuntu-11.0.23+9-post-Ubuntu-1ubuntu120.04.2 generated 12 new + 23 unchanged - 0 fixed = 35 total (was 23)  |
| +1 :green_heart: |  compile  |   9m 11s |  |  the patch passed with JDK Private Build-1.8.0_412-8u412-ga-1~20.04.1-b08  |
| -1 :x: |  javac  |   9m 11s | [/results-compile-javac-root-jdkPrivateBuild-1.8.0_412-8u412-ga-1~20.04.1-b08.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6916/1/artifact/out/results-compile-javac-root-jdkPrivateBuild-1.8.0_412-8u412-ga-1~20.04.1-b08.txt) |  root-jdkPrivateBuild-1.8.0_412-8u412-ga-1~20.04.1-b08 with JDK Private Build-1.8.0_412-8u412-ga-1~20.04.1-b08 generated 21 new + 23 unchanged - 0 fixed = 44 total (was 23)  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| -0 :warning: |  checkstyle  |   0m 42s | [/results-checkstyle-hadoop-common-project_hadoop-common.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6916/1/artifact/out/results-checkstyle-hadoop-common-project_hadoop-common.txt) |  hadoop-common-project/hadoop-common: The patch generated 1 new + 239 unchanged - 0 fixed = 240 total (was 239)  |
| +1 :green_heart: |  mvnsite  |   1m  3s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 38s |  |  the patch passed with JDK Ubuntu-11.0.23+9-post-Ubuntu-1ubuntu120.04.2  |
| +1 :green_heart: |  javadoc  |   0m 36s |  |  the patch passed with JDK Private Build-1.8.0_412-8u412-ga-1~20.04.1-b08  |
| +1 :green_heart: |  spotbugs  |   1m 38s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  24m 19s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| -1 :x: |  unit  |  16m 31s | [/patch-unit-hadoop-common-project_hadoop-common.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6916/1/artifact/out/patch-unit-hadoop-common-project_hadoop-common.txt) |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 43s |  |  The patch does not generate ASF License warnings.  |
|  |   | 132m 58s |  |  |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.net.TestSocketIOWithTimeout |
|   | hadoop.metrics2.source.TestJvmMetrics |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.46 ServerAPI=1.46 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6916/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/6916 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets |
| uname | Linux 34d13a91b041 5.15.0-106-generic #116-Ubuntu SMP Wed Apr 17 09:17:56 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 5ab974d6f9a5c1bc40b5d8429e88baa11c84a94a |
| Default Java | Private Build-1.8.0_412-8u412-ga-1~20.04.1-b08 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.23+9-post-Ubuntu-1ubuntu120.04.2 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_412-8u412-ga-1~20.04.1-b08 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6916/1/testReport/ |
| Max. process+thread count | 1279 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common U: hadoop-common-project/hadoop-common |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6916/1/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

File: hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/ipc/Server.java [hunk 1]
```diff
@@ -2516,13 +2509,15 @@ public int readAndProcess() throws IOException, InterruptedException {
           }
 
           if (!RpcConstants.HEADER.equals(dataLengthBuffer)) {
+            final String hostName = addr == null ? this.hostAddress : addr.getHostName();
```
Comment (shahrs87) [inline hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/ipc/Server.java:2512]: In what case addr would be null? We were not handling `addr == null` edge case before. Wondering what has changed.

**Review by shahrs87 — APPROVED** [review]
+1 The changes looks good to me.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 20s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 2 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  35m 44s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   9m 50s |  |  trunk passed with JDK Ubuntu-11.0.23+9-post-Ubuntu-1ubuntu120.04.2  |
| +1 :green_heart: |  compile  |   9m 17s |  |  trunk passed with JDK Private Build-1.8.0_412-8u412-ga-1~20.04.1-b08  |
| +1 :green_heart: |  checkstyle  |   0m 52s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m  0s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   0m 52s |  |  trunk passed with JDK Ubuntu-11.0.23+9-post-Ubuntu-1ubuntu120.04.2  |
| +1 :green_heart: |  javadoc  |   0m 36s |  |  trunk passed with JDK Private Build-1.8.0_412-8u412-ga-1~20.04.1-b08  |
| +1 :green_heart: |  spotbugs  |   1m 38s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  23m 51s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 32s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   9m 25s |  |  the patch passed with JDK Ubuntu-11.0.23+9-post-Ubuntu-1ubuntu120.04.2  |
| +1 :green_heart: |  javac  |   9m 25s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   9m  6s |  |  the patch passed with JDK Private Build-1.8.0_412-8u412-ga-1~20.04.1-b08  |
| +1 :green_heart: |  javac  |   9m  6s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   0m 44s |  |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   0m 56s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 42s |  |  the patch passed with JDK Ubuntu-11.0.23+9-post-Ubuntu-1ubuntu120.04.2  |
| +1 :green_heart: |  javadoc  |   0m 36s |  |  the patch passed with JDK Private Build-1.8.0_412-8u412-ga-1~20.04.1-b08  |
| +1 :green_heart: |  spotbugs  |   1m 39s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  24m  6s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| -1 :x: |  unit  |  16m 39s | [/patch-unit-hadoop-common-project_hadoop-common.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6916/2/artifact/out/patch-unit-hadoop-common-project_hadoop-common.txt) |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 42s |  |  The patch does not generate ASF License warnings.  |
|  |   | 149m 58s |  |  |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.metrics2.lib.TestMetricsRegistry |
|   | hadoop.net.TestSocketIOWithTimeout |
|   | hadoop.metrics2.source.TestJvmMetrics |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.46 ServerAPI=1.46 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6916/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/6916 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets |
| uname | Linux 30796cf6ad7a 5.15.0-106-generic #116-Ubuntu SMP Wed Apr 17 09:17:56 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 97059adda9f497956038a982c54626aa1d69987d |
| Default Java | Private Build-1.8.0_412-8u412-ga-1~20.04.1-b08 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.23+9-post-Ubuntu-1ubuntu120.04.2 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_412-8u412-ga-1~20.04.1-b08 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6916/2/testReport/ |
| Max. process+thread count | 1716 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common U: hadoop-common-project/hadoop-common |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6916/2/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 29s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 2 new or modified test files.  |
|||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  44m 29s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  17m 24s |  |  trunk passed with JDK Ubuntu-11.0.23+9-post-Ubuntu-1ubuntu120.04.2  |
| +1 :green_heart: |  compile  |  15m 59s |  |  trunk passed with JDK Private Build-1.8.0_412-8u412-ga-1~20.04.1-b08  |
| +1 :green_heart: |  checkstyle  |   1m 19s |  |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   1m 42s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   1m 17s |  |  trunk passed with JDK Ubuntu-11.0.23+9-post-Ubuntu-1ubuntu120.04.2  |
| +1 :green_heart: |  javadoc  |   0m 54s |  |  trunk passed with JDK Private Build-1.8.0_412-8u412-ga-1~20.04.1-b08  |
| +1 :green_heart: |  spotbugs  |   2m 35s |  |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  35m 31s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 55s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  16m 27s |  |  the patch passed with JDK Ubuntu-11.0.23+9-post-Ubuntu-1ubuntu120.04.2  |
| +1 :green_heart: |  javac  |  16m 27s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  15m 51s |  |  the patch passed with JDK Private Build-1.8.0_412-8u412-ga-1~20.04.1-b08  |
| +1 :green_heart: |  javac  |  15m 51s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  checkstyle  |   1m 15s |  |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   1m 38s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   1m 12s |  |  the patch passed with JDK Ubuntu-11.0.23+9-post-Ubuntu-1ubuntu120.04.2  |
| +1 :green_heart: |  javadoc  |   0m 55s |  |  the patch passed with JDK Private Build-1.8.0_412-8u412-ga-1~20.04.1-b08  |
| +1 :green_heart: |  spotbugs  |   2m 44s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  36m 14s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  19m 30s |  |  hadoop-common in the patch passed.  |
| +1 :green_heart: |  asflicense  |   1m  3s |  |  The patch does not generate ASF License warnings.  |
|  |   | 220m  9s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.46 ServerAPI=1.46 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6916/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/6916 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets |
| uname | Linux 2971f14025b3 5.15.0-113-generic #123-Ubuntu SMP Mon Jun 10 08:16:17 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 97059adda9f497956038a982c54626aa1d69987d |
| Default Java | Private Build-1.8.0_412-8u412-ga-1~20.04.1-b08 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.23+9-post-Ubuntu-1ubuntu120.04.2 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_412-8u412-ga-1~20.04.1-b08 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6916/3/testReport/ |
| Max. process+thread count | 1429 (vs. ulimit of 5500) |
| modules | C: hadoop-common-project/hadoop-common U: hadoop-common-project/hadoop-common |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6916/3/console |
| versions | git=2.25.1 maven=3.6.3 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

**Review by zeekling — APPROVED** [review]

**Review by d-c-manning — APPROVED** [review]

**Review by Hexiaoqiao — APPROVED** [review]
LGTM. +1.

**Comment by Hexiaoqiao** [comment]
Committed to trunk. Thanks @virajjasani for your works and @shahrs87 @zeekling @d-c-manning for your reviews!

**Pull request merged**